hazelcast/src/test/java/com/hazelcast/jet/JobListenerTest.java

```diff
@@ -0,0 +1,387 @@
+                    } catch (InterruptedException | ExecutionException e) {
+                        failure = e.getCause();
+                    }
+                    assertNotNull(failure);
+                    sleepSeconds(1);
+                    assertIterableEquals(listener.log,
+                            "Jet: RUNNING -> FAILED (" + failure + ")");
+                });
+    }
+
+    @Test
+    public void testLightListenerDeregistration() {
+        testLightListener(TestSources.itemStream(1, (t, s) -> s),
+                (job, listener) -> {
+                    instance.get().getJet().removeJobStatusListener(listener);
+                    job.cancel();
+                    sleepSeconds(1);
+                    assertTrue(listener.log.isEmpty());
+                });
+    }
+
+    @After
+    public void testListenerDeregistration_onCompletion() throws IllegalAccessException {
+        assertTrue(((Map<?, ?>) jobListeners.get(instance.get().getJet())).isEmpty());
+    }
+
+    /**
+     * @param source A {@link BatchSource BatchSource&lt;Long&gt;} or {@link StreamSource
+     *        StreamSource&lt;Long&gt;} where the first element is {@code 0L}, which is used
+     *        to increment the {@linkplain JobStatusLogger#runCount run count} of the job.
+     */
+    void testListener(JobConfig config, Object source, BiConsumer<Job, JobStatusLogger> test) {
+        Pipeline p = Pipeline.create();
+        int jobId = nextJobId++;
+        (source instanceof BatchSource
+                    ? p.readFrom((BatchSource<Long>) source)
+                    : p.readFrom((StreamSource<Long>) source).withoutTimestamps())
+                .writeTo(Sinks.<Long, Integer, Integer>mapWithUpdating(
+                        MAP_NAME, s -> jobId, (i, s) -> s == 0 ? (i == null ? 0 : i) + 1 : i));
+
+        JetService jet = instance.get().getJet();
+        JobStatusLogger listener = new JobStatusLogger(jobId);
+        Job job = jet.newJob(p, config);
+        jet.addJobStatusListener(set(job.getId()), listener);
+        test.accept(job, listener);
+    }
+
+    void testListener(Object source, BiConsumer<Job, JobStatusLogger> test) {
+        testListener(new JobConfig(), source, test);
+    }
+
+    void testListener(Object source, Consumer<Job> test, String... log) {
+        testListener(source, (job, listener) -> {
+            test.accept(job);
+            assertTailEquals(listener.log, log);
+        });
+    }
+
+    void testLightListener(Object source, BiConsumer<Job, JobStatusLogger> test) {
+        Pipeline p = Pipeline.create();
+        (source instanceof BatchSource
+                    ? p.readFrom((BatchSource<?>) source)
+                    : p.readFrom((StreamSource<?>) source).withoutTimestamps())
+                .writeTo(Sinks.noop());
+
+        JetService jet = instance.get().getJet();
+        JobStatusLogger listener = new JobStatusLogger(-1);
+        Job job = jet.newLightJob(p);
+        sleepSeconds(1);
```

Comment:
This can be removed by registering the listener via `job.addStatusListener(listener)`, but it will
make `testListenerDeregistration_onCompletion` useless since the Jet service is bypassed.